lib/tasks_private/instance_types.rake

```diff
@@ -1,215 +1,277 @@
+      end.to_h
 
-  puts <<EOT
-Created ./instance_types.rb with #{new_count} new, #{deprecated_count} newly deprecated and #{discontinued_count} new discontinued instances.
-Now copy it to the existing one:
+      [types_data, collecting_warnings, parsing_warnings]
+    end
+
+    ## consider previous data
+
+    old_types_data = YAML.load_file(out_file)
+    types_list = types_list | old_types_data.keys | types_data.keys
+
+    default_type = old_types_data.find { |_, data| data[:default] }.first
+    discontinued_types |= (old_types_data.keys - types_data.keys)
+    types_data.merge!(old_types_data.slice(*discontinued_types))
+
+    ## postprocess
+
+    types_data.sort_by! do |instance_type, instance_data|
+      instance_data.except!(:default, :deprecated, :discontinued, :disabled)
+      instance_data[:default] = true if instance_type == default_type
+      if discontinued_types.include?(instance_type) || !instance_data[:current_version]
+        instance_data[:discontinued] = true
+        instance_data[:disabled] = true
+      end
+      if instance_data[:current_version] && !instance_data[:current_generation]
+        instance_data[:deprecated] = true
+      end
+      types_list.index(instance_type) || 1_000_000
+    end
+
+    ## show warnings
+
+    unknown_types = types_list - (types_data.keys & types_list)
+
+    unless collecting_warnings.empty?
+      info = collecting_warnings.transform_keys(&:first)
+      info.each_value do |instance_data|
+        instance_data.transform_values!(&:to_a)
+      end
+      info = info.sort_by! do |instance_type, _|
+        types_list.index(instance_type) || 1_000_000
+      end
+      log_data('Attention! Contradictory products data:', info)
+    end
+    log_data('Attention! Unforeseen values format:', parsing_warnings)
+    log_data('Attention! Undeclared or unknown instance types:', unknown_types)
+
+    ## generate update report
+
+    isolated do
+      report_date = Time.now.utc.to_datetime
+      report_name = "instance_types_diff_#{report_date.strftime('%Y-%m-%d-%H-%M-%S')}.html"
+      report_file = Rails.root.join('tmp', report_name)
+
+      old = old_types_data.deep_dup
+      new = types_data.deep_dup
+
+      report = +''
+      report << '<html>'
+      report << '<head>'
+      report << '<meta charset="utf-8">'
+      report << '<style>'
+      report << 'table{width:100%}'
+      report << 'th,td{width:36%}'
+      report << 'th,td:first-child{width:28%}'
+      report << 'th,td{border:1px dotted gray}'
+      report << '.added{background-color:rgb(223,240,216)}'
+      report << '.removed{background-color:rgb(242,222,222)}'
+      report << '.changed{background-color:rgb(252,248,227);}'
+      report << '</style>'
+      report << '</head>'
+      report << '<body>'
+      report << '<h1>Generated: <script>'
+      report << "document.write(new Date('#{report_date.iso8601}').toString())"
+      report << '</script></h1>'
```

Comment:
I don't understand the point of this report

Reply:
It shows what attributes was changed - for control. I've used it to control data consistency before my PR and after.
However, this report doesn't go anywhere besides local `tmp` dir.

Reply:
Ok, yeah, then IMO it's completely unnecessary...`git diff` should be more than sufficient for those purposes.

Reply:
`git diff` can't guarantee you consistent experience.
You can see it here for example: https://github.com/AlexanderZagaynov/manageiq-providers-amazon/compare/3a9a45211d9374470c5051b87e0b1187cdcc30ac...a92033eb4852eda82bb5a8263305e63067380a0e#diff-d8dde9392fe3a31bd3f5580c14571a1cL1681
(I'm attaching a screenshot of this)
![screenshot from 2018-07-03 16-53-27](https://user-images.githubusercontent.com/7778414/42227473-e9b341fe-7ee1-11e8-9339-77bfddf5302b.png)

Reply:
The link doesn't work...do you have another link with an example?  Wanted to see if `git diff --patience` made a difference at all